### PR TITLE
Add Result.Enum.map_grouped_by_tag/2

### PR DESCRIPTION
**Pull request by giddie**

### CODE REVIEW

**Review by tshakah — APPROVED** [review]
This looks great - exactly what I needed the other day

File: lib/ok_then/result/enum.ex [hunk 1]
```diff
@@ -79,6 +79,74 @@ defmodule OkThen.Result.Enum do
...
+    |> group_by_tag()
+    |> Enum.flat_map(fn {tag, values} ->
+      filter_function.(tag, values)
+      |> case do
+        list when is_list(list) ->
+          Enum.map(list, &{tag, &1})
+
+        other ->
+          raise ArgumentError,
+                "Expected filter_function clause for tag #{Kernel.inspect(tag)} " <>
```
Comment (tshakah) [inline lib/ok_then/result/enum.ex:144]: The formatting is a little janky here, but it's technically valid so 🤷🏽

Comment (giddie) [reply lib/ok_then/result/enum.ex:144]: Yeah, it's not brilliant. Can you think of a way to improve it?

Comment (tshakah) [reply lib/ok_then/result/enum.ex:144]: In Rust I would use `format!`, but I don't think Elixir has an equivalent

File: lib/ok_then/result/enum.ex [hunk 2]
```diff
@@ -79,6 +79,74 @@ defmodule OkThen.Result.Enum do
...
+                  "to return a list, but got: #{Kernel.inspect(other)}"
```
Comment (tshakah) [inline lib/ok_then/result/enum.ex:145]:
```suggestion
          raise ArgumentError,
                "Expected filter_function clause for tag #{Kernel.inspect(tag)} \
                to return a list, but got: #{Kernel.inspect(other)}"
```

File: lib/ok_then/result/enum.ex (hunk 1 above)
Comment (tshakah) [reply lib/ok_then/result/enum.ex:144]: I can't think of a better way without introducing a different type of janky

**Comment by giddie** [comment]
Merged manually

**Pull request closed without merging**